packages/jupyter-ai/jupyter_ai/tools/models.py

```diff
@@ -0,0 +1,232 @@
+        if not self.name:
+            if hasattr(self.callable, "__name__") and self.callable.__name__:
+                self.name = self.callable.__name__
+            else:
+                raise ValueError("Unable to extract name from callable")
+
+        if not self.description:
+            self.description = get_doc_description(self.callable)
+
+        return self
+
+    def __eq__(self, other):
+        if not isinstance(other, Tool):
+            return False
+        return self.name == other.name
+
+    def __hash__(self):
+        return hash(self.name)
+
+
+class ToolSet(set):
+    """A collection of tools that enforces unique tool names.
+
+    Ensures that tools with duplicate names cannot be added to the same collection.
+
+    Example:
+        >>> toolset = ToolSet()
+        >>> tool1 = Tool(callable=lambda: None, name="test")
+        >>> tool2 = Tool(callable=lambda: None, name="test")  # Same name
+        >>> toolset.add(tool1)
+        >>> toolset.add(tool2)  # Raises ValueError
+        Traceback (most recent call last):
+        ...
+        ValueError: Tool with name 'test' already exists in the set
+    """
+
+    def add(self, item):
+        if item in self:
+            raise ValueError(f"Tool with name '{item.name}' already exists in the set")
+        super().add(item)
+
+
+class Toolkit(BaseModel):
+    """A collection of tools with capability-based filtering.
+
+    A Toolkit groups related tools together and provides methods to add tools
+    and find tools based on their capability flags.
+
+    Attributes:
+        name: The toolkit's name.
+        description: Optional description of the toolkit.
+        tools: The set of tools in this toolkit.
+
+    Example:
+        >>> def read_file(path: str):
+        ...     '''Read a file from disk.'''
+        ...     with open(path) as f:
+        ...         return f.read()
+        >>> def write_file(path: str, content: str):
+        ...     '''Write content to a file.'''
+        ...     with open(path, 'w') as f:
+        ...         f.write(content)
+        >>> toolkit = Toolkit(name="FileTools", description="File operations")
+        >>> read_tool = Tool(callable=read_file, read=True)
+        >>> write_tool = Tool(callable=write_file, write=True)
+        >>> toolkit.add_tool(read_tool)
+        >>> toolkit.add_tool(write_tool)
+        >>> len(toolkit.tools)
+        2
+        >>> read_tools = toolkit.find_tools(read=True)
+        >>> len(read_tools)
+        1
+    """
+
+    name: str
+    description: Optional[str] = None
+    tools: ToolSet = Field(default_factory=ToolSet)
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def add_tool(self, tool: Tool):
+        """Add a tool to this toolkit.
+
+        Args:
+            tool: The tool to add to the toolkit.
+
+        Raises:
+            ValueError: If a tool with the same name already exists in the toolkit.
+        """
+        self.tools.add(tool)
+
+    def find_tools(
```

Comment:
Because this is doing a direct lookup and not a search, I think get_tools would be a better reflection of what this is doing.